### ISSUE TITLE
Move canvas when hitting canvas edge while dragging an item

### ISSUE THREAD

**Issue by emielwit** [wbkd/react-flow#460]
Hey,

I have a sidebar with elements where I can drag items on the canvas. I'm curious if there is a function that I can get from the store or something to move the canvas when I want to drag an item on a tree that is not visible. So when I hit the edge of the canvas the canvas is moving to that side. For example, I want to drop my item on one of the elements that are not visible (see image). 

![image](https://user-images.githubusercontent.com/15437007/91165595-3631ad00-e6d1-11ea-90b4-8af7a2626a11.png)

**Comment by moklick** [comment]
Hey,

we have a similar issue #392 where a user wants to move the pane when a node gets dragged out of the viewport.

You could try to use the `setInitTransform` action. This action also updates the internal d3-zoom state and not only the `transform` store state.

**Comment by afallon02** [comment]
This would be an amazing feature - +1 for it's implementation

**Comment by emielwit** [comment]
Thanks, I will take a look if I can find a solution for this. I will close the issue for now :)

**Comment by moklick** [comment]
Let us know of you do :)

**Comment by emielwit** [comment]
I found a solution, FYI I'm using `react-dnd` for dragging items from a sidebar on the canvas.

I made a component that I can place on top of the canvas. I render this component only when I start dragging. When I hover over one of the custom components its starts to move the canvas. I use the state, store, and useEffect to trigger the canvas to move. I still need to give the `MoveCanvas` component a direction. So when I render this component 4 times for (top, bottom, left, right) it knows what direction it needs to go.

```tsx
import React, { FC, useState, useEffect } from 'react';
import { useStoreState, useStoreActions } from 'react-flow-renderer';
import styled from 'styled-components';
import { useDrop } from 'react-dnd';
import { DragTypes } from '../nodes/types';

const StyledDiv = styled.div`
  position: absolute;
  bottom: 2rem;
  height: 100px;
  width: 100%;
`;

const MoveCanvas: FC<{}> = () => {
  const [onDragEnter, setOnDragEnter] = useState<boolean>(false);
  const { setInitTransform } = useStoreActions(actions => actions);
  const [x, y, zoomLevel] = useStoreState(store => {
    return store.transform;
  });
  const [, drop] = useDrop({
    accept: DragTypes.STEP,
    drop: () => {
      setOnDragEnter(false);
    },
  });

  useEffect(() => {
    if (onDragEnter) {
      setInitTransform({ x, y: y - 6, k: zoomLevel });
    }
  }, [onDragEnter, setInitTransform, x, y, zoomLevel]);

  return (
    <StyledDiv
      ref={drop}
      onDragEnter={(): void => {
        setOnDragEnter(true);
      }}
      onDragLeave={(): void => {
        setOnDragEnter(false);
      }}
    />
  );
};

export default MoveCanvas;
```

**Comment by afallon02** [comment]
@emielwit that is a really cool solution - thank you so much for sharing. Also - I hadn't heard about react-dnd and that looks like a really cool library so thanks for sharing.

**Comment by moklick** [comment]
I am glad you found a solution you could build around react flow :) I will have a look how you solved it, thanks!

**Comment by moklick** [comment]
This feature is available since v11.5.0.
